daft/filesystem.py

```diff
@@ -71,18 +105,226 @@ def get_filesystem(protocol: str, **kwargs) -> fsspec.AbstractFileSystem:
+        )
+
+    # Canonical protocol shared by all paths.
+    protocol = next(iter(canonicalized_protocols))
+
+    if filesystem is None:
+        # Try to get filesystem from protocol -> fs cache.
+        filesystem = _get_fs_from_cache(protocol)
+    elif isinstance(filesystem, fsspec.AbstractFileSystem):
+        # Wrap fsspec filesystems so they are valid pyarrow filesystems.
+        filesystem = PyFileSystem(FSSpecHandler(filesystem))
+
+    # Resolve path and filesystem for the first path.
+    # We use this first resolved filesystem for validation on all other paths.
+    resolved_path, resolved_filesystem = _resolve_path_and_filesystem(paths[0], filesystem)
+
+    if filesystem is None:
+        filesystem = resolved_filesystem
+        # Put resolved filesystem in cache under these paths' canonical protocol.
+        _put_fs_in_cache(protocol, filesystem)
+
+    # filesystem should be a non-None pyarrow FileSystem at this point, either
+    # user-provided, taken from the cache, or inferred from the first path.
+    assert filesystem is not None and isinstance(filesystem, FileSystem)
+
+    # Resolve all other paths and validate with the user-provided/cached/inferred filesystem.
+    resolved_paths = [resolved_path]
+    for path in paths[1:]:
+        resolved_path, _ = _resolve_path_and_filesystem(path, filesystem)
+        resolved_paths.append(resolved_path)
+
+    return resolved_paths, filesystem
+
+
+def _resolve_path_and_filesystem(
+    path: str,
+    filesystem: FileSystem | fsspec.AbstractFileSystem | None,
+) -> tuple[str, FileSystem]:
+    """
+    Resolves and normalizes the provided path, infers a filesystem from the
+    path, and ensures that the inferred filesystem is compatible with the passed
+    filesystem, if provided.
+
+    Args:
+        path: A single file/directory path.
+        filesystem: The filesystem implementation that should be used for
+            reading these files. If None, a filesystem will be inferred. If not
+            None, the provided filesystem will still be validated against the
+            filesystem inferred from the provided path to ensure compatibility.
+    """
+    from loguru import logger
+
+    # Use pyarrow utility to resolve filesystem and this particular path.
+    # If a non-None filesystem is provided to this utility, it will ensure that
+    # it is compatible with the provided path.
+    # A non-None filesystem will be provided if:
+    #  - a user-provided filesystem was passed to _resolve_paths_and_filesystem.
+    #  - a filesystem for the paths' protocol exists in the protocol -> fs cache.
+    #  - a filesystem was resolved for a previous path; i.e., filesystem is
+    #    guaranteed to be non-None for all but the first path.
+    try:
+        resolved_filesystem, resolved_path = _resolve_filesystem_and_path(path, filesystem)
+    except pa.lib.ArrowInvalid as e:
+        if "Cannot parse URI" in str(e):
```

Comment:
Would it be better to always sanitize the URLs at the start of the function, instead of only on retries?